hazelcast/src/main/java/com/hazelcast/internal/metrics/managementcenter/MetricsCompressor.java

```diff
@@ -0,0 +1,218 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.metrics.managementcenter;
+
+import com.hazelcast.nio.Bits;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.InflaterInputStream;
+
+import static com.hazelcast.internal.metrics.MetricsUtil.escapeMetricNamePart;
+import static java.lang.Math.multiplyExact;
+
+/**
+ * Helper class for compressing metrics into a byte[] blob to be read by
+ * various consumers. The consumer itself is passed in as a parameter.
+ * <p>
+ * Most efficient when the metrics to be compressed are fed in sorted by
+ * name. It will work even if this condition is violated, but it won't
+ * nearly be as efficient.
+ * <p>
+ * Before compressing it also converts legacy metric names to
+ * {@code [metric=<oldName>]}.
+ * <p>
+ * The utility is optimized to repeatedly compress similar set of metrics: it
+ * reuses the buffer and keeps it about 10% bigger than what was needed the
+ * last time.
+ */
+public class MetricsCompressor {
+
+    @SuppressWarnings({"checkstyle:MagicNumber", "checkstyle:TrailingComment"})
+    private static final int INITIAL_BUFFER_SIZE = 2 << 11; // 2kB
+    private static final int SIZE_FACTOR_NUMERATOR = 11;
+    private static final int SIZE_FACTOR_DENOMINATOR = 10;
+
+    private static final int BITS_IN_BYTE = 8;
+    private static final int BYTE_MASK = 0xff;
+
+    private static final short SHORT_BITS = BITS_IN_BYTE * Bits.SHORT_SIZE_IN_BYTES;
+    // required precision after the decimal point for doubles
+    private static final int CONVERSION_PRECISION = 4;
+    // coefficient for converting doubles to long
+    private static final double DOUBLE_TO_LONG = Math.pow(10, CONVERSION_PRECISION);
+
+    private static final short BINARY_FORMAT_VERSION = 1;
+
+    private DataOutputStream dos;
+    private MorePublicByteArrayOutputStream baos = new MorePublicByteArrayOutputStream(INITIAL_BUFFER_SIZE);
+    private String lastName;
+    private int count;
+
+    public MetricsCompressor() {
+        reset(INITIAL_BUFFER_SIZE);
+    }
+
+    public void addLong(String name, long value) {
+        try {
+            writeName(name);
+            dos.writeLong(value);
+        } catch (IOException e) {
+            // should never be thrown
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void addDouble(String name, double value) {
+        try {
+            writeName(name);
+            // convert to long with specified precision
+            dos.writeLong(Math.round(value * DOUBLE_TO_LONG));
```

Comment:
A really minor consideration. Looks like we're loosing information about the type of metric here. It's fine, but it creates additional coupling: MC has to know which metrics have `double` type. Maybe, it could be avoided by adding a special tag for such metrics (say, `type=double`) or an additional byte header with the type id. Then `decompressingIterator` could be restoring those `double`s (with some precision loss) and saving them into `Metric` objects.

Reply:
It only supports `long` type of metric, and the only double metrics are things like cpu usage where you don't care about precision to that degree.

Reply:
It's not about precision loss, it's about lost value type. It's absolutely fine that the compression is lossy. But it would be nice if the iterator could recover types when data points are uncompressed. In this case, it'll be easier to support additional data point types in the future (see my considerations above).

And as I mentioned above, MC will work fine with those lost types, as it's aware of metrics and can convert metrics with double values based on their names.

Reply:
I agree we need to carry the types. I'd address this is in a separate PR so that integration with MC side gets unblocked.